builder/optimizations.cc

```diff
@@ -843,6 +843,91 @@ stmt alias_in_place(const stmt& s, const std::vector<buffer_expr_ptr>& outputs)
   return in_place_aliaser(outputs).mutate(s);
 }
 
+namespace {
+
+template <typename T>
+bool match(span<const T> a, span<const T> b) {
+  if (a.size() != b.size()) return false;
+  for (std::size_t i = 0; i < a.size(); ++i) {
+    if (!match(a[i], b[i])) return false;
+  }
+  return true;
+}
+
+class sibling_fuser : public stmt_mutator {
+  static bool can_fuse(const allocate* a, const allocate* b) {
+    return a->storage == b->storage && match(a->elem_size, b->elem_size) && match<dim_expr>(a->dims, b->dims);
+  }
+  static bool can_fuse(const make_buffer* a, const make_buffer* b) {
+    return match(a->base, b->base) && match(a->elem_size, b->elem_size) && match<dim_expr>(a->dims, b->dims);
+  }
+  static bool can_fuse(const crop_dim* a, const crop_dim* b) {
+    return a->src == b->src && a->dim == b->dim && match(a->bounds, b->bounds);
+  }
+  static bool can_fuse(const crop_buffer* a, const crop_buffer* b) {
+    return a->src == b->src && match<interval_expr>(a->bounds, b->bounds);
+  }
+  static bool can_fuse(const slice_dim* a, const slice_dim* b) {
+    return a->src == b->src && a->dim == b->dim && match(a->at, b->at);
+  }
+  static bool can_fuse(const slice_buffer* a, const slice_buffer* b) {
+    return a->src == b->src && match<expr>(a->at, b->at);
+  }
+  static bool can_fuse(const transpose* a, const transpose* b) { return a->src == b->src && a->dims == b->dims; }
+
+  template <typename T>
+  static bool fuse(const T* a, const T* b, stmt& result) {
+    if (!a || !b || !can_fuse(a, b)) return false;
+
+    stmt body = block::make({a->body, substitute(b->body, b->sym, a->sym)});
+    result = clone_with(a, std::move(body));
+    return true;
+  }
+
+  static bool fuse(stmt& a, const stmt& b) {
+    return fuse(a.as<allocate>(), b.as<allocate>(), a) || fuse(a.as<make_buffer>(), b.as<make_buffer>(), a) ||
+           fuse(a.as<crop_dim>(), b.as<crop_dim>(), a) || fuse(a.as<crop_buffer>(), b.as<crop_buffer>(), a) ||
```

Comment:
I think this needs a comment to explaining that crop_buffer/crop_dim/etc can only be fused if they have the same src